tests/unit/parallel/algorithms/destroy_tests.hpp

```diff
@@ -0,0 +1,358 @@
+//  Copyright (c) 2014-2017 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_PARALLEL_TEST_UNINIT_COPY_MAY28_15_1344)
+#define HPX_PARALLEL_TEST_UNINIT_COPY_MAY28_15_1344
+
+#include <hpx/hpx_init.hpp>
+#include <hpx/hpx.hpp>
+#include <hpx/include/parallel_executors.hpp>
+#include <hpx/include/parallel_destroy.hpp>
+#include <hpx/util/lightweight_test.hpp>
+
+#include <boost/atomic.hpp>
+#include <boost/range/functions.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <utility>
+
+#include "test_utils.hpp"
+
+struct destructable
+{
+    destructable()
+      : value_(0)
+    {}
+
+    ~destructable()
+    {
+        std::memset(&value_, 0xcd, sizeof(value_));
+    }
+
+    std::uint32_t value_;
+};
+
+std::size_t const data_size = 10007;
+
+////////////////////////////////////////////////////////////////////////////
+template <typename ExPolicy, typename IteratorTag>
+void test_destroy(ExPolicy && policy, IteratorTag)
+{
+    static_assert(
+        hpx::parallel::execution::is_execution_policy<ExPolicy>::value,
+        "hpx::parallel::execution::is_execution_policy<ExPolicy>::value");
+
+    typedef destructable* base_iterator;
+    typedef test::test_iterator<base_iterator, IteratorTag> iterator;
+
+    destructable* p = (destructable*)std::malloc(
+        data_size * sizeof(destructable));
+
+    // value-initialize data in array
+    std::for_each(
+        p, p + data_size,
+        [](destructable& d)
+        {
+            ::new (static_cast<void*>(std::addressof(d))) destructable;
+        });
+
+    hpx::parallel::destroy(
+        std::forward<ExPolicy>(policy),
+        iterator(p), iterator(p + data_size));
+
+    std::size_t count = 0;
+    std::for_each(p, p + data_size,
+        [&count](destructable v1)
+        {
+            HPX_TEST_EQ(v1.value_, (std::uint32_t)0xcdcdcdcd);
```

Comment:
It is... Do you have a better idea how to check whether an object actually was destroyed? We could perhaps simply use a counter to count invocations. Not sure.